src/dimension.c

```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
+				 errmsg("Invalid number of partitions")));
+
+	num_slices = num_slices_arg & 0xffff;
+
+	dimension_update(fcinfo, table_relid, colname, DIMENSION_TYPE_CLOSED, NULL, &num_slices);
+
+	PG_RETURN_VOID();
+}
+
+TS_FUNCTION_INFO_V1(dimension_set_interval);
+
+Datum
+dimension_set_interval(PG_FUNCTION_ARGS)
+{
+	Oid			table_relid = PG_GETARG_OID(0);
+	Datum		interval = PG_GETARG_DATUM(1);
+	Name		colname = PG_ARGISNULL(2) ? NULL : PG_GETARG_NAME(2);
+
+	hypertable_permissions_check(table_relid, GetUserId());
+
+	if (PG_ARGISNULL(1))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid interval: an explicit interval must be specified")));
+
+	dimension_update(fcinfo, table_relid, colname, DIMENSION_TYPE_OPEN, &interval, NULL);
+
+	PG_RETURN_VOID();
+}
+
+/*
+ * Dimension information used to validate, create and update dimensions.
+ */
+typedef struct DimensionInfo
+{
+	Oid			table_relid;
+	Name		colname;
+	Oid			coltype;
+	DimensionType type;
+	Datum		interval_datum;
+	Oid			interval_type;	/* Type of the interval datum */
+	int64		interval;
+	int32		num_slices;
+	regproc		partitioning_func;
+	bool		if_not_exists;
+	bool		skip;
+	bool		set_not_null;
+	bool		num_slices_is_set;
+	bool		show_warnings;
+	Hypertable *ht;
+} DimensionInfo;
+
+#define DIMENSION_INFO_IS_SET(di)										\
+	(OidIsValid((di)->table_relid) &&									\
+	 (di)->colname != NULL &&											\
+	 ((di)->num_slices_is_set || OidIsValid((di)->interval_datum)))
+
+static void
+dimension_validate_info(DimensionInfo *info)
+{
+	Dimension  *dim;
+	HeapTuple	tuple;
+	Datum		datum;
+	bool		isnull = false;
+	bool		not_null_is_set;
+
+	if (!DIMENSION_INFO_IS_SET(info))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid dimension info")));
+
+	/* Check that the column exists and get its NOT NULL status */
+	tuple = SearchSysCacheAttName(info->table_relid, NameStr(*info->colname));
+
+	if (!HeapTupleIsValid(tuple))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_COLUMN),
+				 errmsg("column \"%s\" does not exist",
+						NameStr(*info->colname))));
+
+	datum = SysCacheGetAttr(ATTNAME,
+							tuple,
+							Anum_pg_attribute_atttypid,
+							&isnull);
+	Assert(!isnull);
+
+	info->coltype = DatumGetObjectId(datum);
+
+	datum = SysCacheGetAttr(ATTNAME,
+							tuple,
+							Anum_pg_attribute_attnotnull,
+							&isnull);
+	Assert(!isnull);
+
+	not_null_is_set = DatumGetBool(datum);
+
+	ReleaseSysCache(tuple);
+
+	if (NULL != info->ht)
+	{
+		/* Check if the dimension already exists */
+		dim = hyperspace_get_dimension_by_name(info->ht->space,
+											   DIMENSION_TYPE_ANY,
+											   NameStr(*info->colname));
+
+		if (NULL != dim)
+		{
+			if (!info->if_not_exists)
+				ereport(ERROR,
+						(errcode(ERRCODE_IO_DUPLICATE_DIMENSION),
+						 errmsg("column \"%s\" is already a dimension",
+								NameStr(*info->colname))));
+
+			info->skip = true;
+
+			ereport(NOTICE,
```

Comment:
do NOTICES come with error codes?

Reply:
Good question... Probably not. How would you check the error code if we do not return here and/or have multiple notices?